common/archiver/filestore/historyArchiver.go

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package filestore
+
+import (
+	"context"
+	"errors"
+	"path"
+	"strings"
+
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/archiver"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/persistence"
+)
+
+const (
+	fileStoreScheme = "file://"
+
+	archiveNonRetriableErrorMsg = "Archive method encountered an non-retriable error."
+
+	errInvalidURI               = "archive URI is invalid"
+	errInvalidRequest           = "archive request is invalid"
+	errConstructHistoryIterator = "failed to construct history iterator"
+	errReadHistory              = "failed to read history batches"
+	errHistoryMutated           = "history was mutated"
+	errEncodeHistory            = "failed to encode history batches"
+	errMakeDirectory            = "failed to make directory"
+	errWriteFile                = "failed to write history to file"
+)
+
+var (
+	errContextTimeout           = errors.New("activity aborted because context timed out")
+	errInvalidGetHistoryRequest = &shared.BadRequestError{Message: "Get archived history request is invalid"}
+	errGetHistoryTokenCorrupted = &shared.BadRequestError{Message: "Next page token is corrupted."}
+	errHistoryNotExist          = &shared.BadRequestError{Message: "Requested workflow history does not exist."}
+)
+
+type (
+	// HistoryArchiverConfig configs the filestore implementation of archiver.HistoryArchiver interface
+	HistoryArchiverConfig struct {
+		*archiver.HistoryIteratorConfig
+	}
+
+	historyArchiver struct {
```

Comment:
I think the state for all of these should only contain two fields: providerContainer and config. Basically the provider has a bag of stuff it knows it uses to create different archivers. Each archiver also has a config. Those should be the only two pieces of state needed for archiver implementations.